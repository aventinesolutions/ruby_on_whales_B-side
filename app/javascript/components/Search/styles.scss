.whiskeys-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.search-container {
  border-radius: 1em;
  border: 0.5em solid cornflowerblue;
  box-shadow: 0.2em 0.2em gray;
  background-color: white;
  max-height: 5em;
  max-width: 18em;
  padding: 2em;
  margin: 1em;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  h2 {
    margin-right: 1em;
  }
  input {
    font-size: 1.2em;
    margin-top: 0.5em;
    padding: 0.5em;
    max-height: 2em;
    max-width: 8em;
  }
}
