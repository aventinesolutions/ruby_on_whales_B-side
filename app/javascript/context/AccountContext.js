import { createContext } from 'react';

const Root = document ? document.getElementById('root') : null;

const AccountContext = createContext(
  { account_id: Root ? Root.getAttribute('data-account-id') : null }
);

AccountContext.displayName = 'AccountContext';

export default AccountContext;
