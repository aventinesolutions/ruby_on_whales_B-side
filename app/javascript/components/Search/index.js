import React, { useState } from 'react';
import gql from 'graphql-tag';
import Whiskeys from '../Whiskeys';
import { AccountContext } from "../../context/AccountContext";
import './styles.scss';

const Search = () => {
  const [text, setText] = useState("");

  return (
    <AccountContext.Consumer>
      {
        ({ account_id }) =>
          <div className="whiskeys-container">
            <div className="search-container">
              <h2>Search</h2>
              <input id="search"
                type="text"
                value={text}
                onChange={event => setText(event.target.value)} />
            </div>
            { text.length > 2 ? <Whiskeys query={gql`
              {
                 search(filter: {accountId: "${account_id}", text: "${text}"}) {
                  id
                  title
                  description
                  price
                  photoUrl
                  ratings {
                    id
                    accountId
                    quality
                    stars
                  }
                }
              }
          `} /> : null
            }
          </div>
      }
    </AccountContext.Consumer>
  );
};

Search.displayName = 'Search';

export default Search;
