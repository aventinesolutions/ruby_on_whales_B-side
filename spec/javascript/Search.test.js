import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';

const fake = require('faker');
const wait = require('waait');

import Search from '../../app/javascript/components/Search';
import AccountContext from '../../app/javascript/context/AccountContext';
import { MockedProvider } from "@apollo/react-testing";

describe('shallow rendering Search component', () => {
  let output;

  beforeEach(() => {
    output = shallow(
      <AccountContext.Provider value={{ account_id: fake.random.uuid() }}>
        <Search />
      </AccountContext.Provider>
    );
  });

  it('renders a Search component', () => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});

describe('reacts to changes in Search filters', () => {
  let component;
  let queryCalled;
  let mocks;

  beforeEach(() => {
    queryCalled = false;
    mocks = {};
    component = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AccountContext.Provider value={{ account_id: fake.random.uuid() }}>
          <Search />
        </AccountContext.Provider>
      </MockedProvider>
    );
  });

  it('reacts to changes in filters', async () => {
    expect(component.find('#search').props().value).toEqual("");
    await act(async () => {
      component.find('#search').props().value = "casks";
    });
    await wait(0);
    expect(component.find('#search').props().value).toEqual("casks");
    expect(queryCalled).toBeTruthy();
  });
});
