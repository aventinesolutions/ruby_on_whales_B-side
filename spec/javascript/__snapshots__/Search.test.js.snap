// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow rendering Search component renders a Search component 1`] = `<Search />`;
